pkg/frontend/querymiddleware/roundtrip.go

```diff
@@ -253,17 +260,30 @@ func newQueryTripperware(
 		activeNativeHistogramMetrics := next
 		labels := next
 
-		// Inject the cardinality and labels query cache roundtripper only if the query results cache is enabled.
-		if cfg.CacheResults {
-			cardinality = newCardinalityQueryCacheRoundTripper(c, cacheKeyGenerator, limits, cardinality, log, registerer)
-			labels = newLabelsQueryCacheRoundTripper(c, cacheKeyGenerator, limits, labels, log, registerer)
-		}
-
 		if cfg.ShardActiveSeriesQueries {
 			activeSeries = newShardActiveSeriesMiddleware(activeSeries, cfg.UseActiveSeriesDecoder, limits, log)
 			activeNativeHistogramMetrics = newShardActiveNativeHistogramMetricsMiddleware(activeNativeHistogramMetrics, limits, log)
 		}
 
+		// Enforce read consistency after caching.
```

Comment:
this made me realize that strong read consistency can still use cached responses, which might not always be expected. Should we skip caching when strong read consistency is requested? This feels outside of the scope of this PR though, so it's mostly a discussion point

Reply:
In my opinion cache comes first. It's like work in Mimir classic architecture too, and I would just keep parity. If you want to bypass, then you should pass `Cache-Control: no-store`. But I'm open to discuss it.

Reply:
I guess it comes down to what we promise with strong read consistency. If we want to give strong read consistency on all samples, then we should disable cache for strong read consistency queries.

Otherwise, if we keep caching, we'd potentially not guarantee strong read consistency on out of order samples. It's hard to explain that was we promise is "the latest data except if the query was ran before and there's older data points which arrived in the last 10 minutes".

We don't want to promote this feature, but I still feel like a less surprising behaviour is better. Even if that means worse query performance (which also happens to create a disincentive to use strong consistency when you don't need it).

Reply:
I see your point and I don't disagree. But in the ingest storage the strong read consistency has been introduced to have a way to run queries on par with classic architecture, and that's just what I'm doing here.